runelite-client/src/main/java/net/runelite/client/plugins/kingdomofmiscellania/KingdomPlugin.java

```diff
@@ -122,20 +202,92 @@ private void removeKingdomInfobox()
 		}
 	}
 
+	private int estimateCoffer(long lastChangedEpoch, int lastCoffer)
+	{
+		int daysSince = (int) Duration.between(Instant.ofEpochSecond(lastChangedEpoch), Instant.now()).toDays();
+		int maxDailyWithdrawal = isRoyalTroubleCompleted() ? MAX_WITHDRAWAL_ROYAL_TROUBLE : MAX_WITHDRAWAL_BASE;
+		int maxDailyThreshold = maxDailyWithdrawal * 10;
+
+		for (int i = 0; i < daysSince; i++)
+		{
+			lastCoffer -= (lastCoffer > maxDailyThreshold) ? maxDailyWithdrawal : lastCoffer / 10;
+		}
+		return lastCoffer;
+	}
+
+	private int estimateFavor(long lastChangedEpoch, int lastFavor)
+	{
+		int daysSince = (int) Duration.between(Instant.ofEpochSecond(lastChangedEpoch), Instant.now()).toDays();
+		float dailyPercentage = isRoyalTroubleCompleted() ? FAVOR_DECREMENT_ROYAL_TROUBLE : FAVOR_DECREMENT_BASE;
+
+		lastFavor -= (int) (daysSince * dailyPercentage * MAX_FAVOR);
+		return Math.max(lastFavor, 0);
+	}
+
 	private boolean isInKingdom()
 	{
 		return client.getLocalPlayer() != null
 			&& KINGDOM_REGION.contains(client.getLocalPlayer().getWorldLocation().getRegionID());
 	}
 
-	private boolean hasCompletedQuest()
+	private boolean isThroneOfMiscellaniaCompleted()
 	{
 		return client.getVar(VarPlayer.THRONE_OF_MISCELLANIA) > 0;
 	}
 
+	private boolean isRoyalTroubleCompleted()
+	{
+		return Quest.ROYAL_TROUBLE.getState(client) == QuestState.FINISHED;
+	}
+
 	static int getFavorPercent(int favor)
 	{
-		return (favor * 100) / 127;
+		return (favor * 100) / MAX_FAVOR;
+	}
+
+	private void sendChatMessage(String chatMessage)
+	{
+		final String message = new ChatMessageBuilder()
+			.append(ChatColorType.HIGHLIGHT)
+			.append(chatMessage)
+			.build();
+
+		chatMessageManager.queue(
+			QueuedMessage.builder()
+				.type(ChatMessageType.CONSOLE)
+				.runeLiteFormattedMessage(message)
+				.build());
 	}
 
+	private Instant getLastChanged()
+	{
+		return configManager.getRSProfileConfiguration(KingdomConfig.CONFIG_GROUP_NAME, CONFIG_LAST_CHANGED_KEY, Instant.class);
+	}
+
+	private void setLastChanged(Instant lastChanged)
+	{
+		configManager.setRSProfileConfiguration(KingdomConfig.CONFIG_GROUP_NAME, CONFIG_LAST_CHANGED_KEY, lastChanged);
+	}
+
+	int getCoffer()
+	{
+		Integer coffer = configManager.getRSProfileConfiguration(KingdomConfig.CONFIG_GROUP_NAME, CONFIG_COFFER_KEY, int.class);
+		return coffer == null ? 0 : coffer;
+	}
```

Comment:
It strikes me that these should be nullable as we would not want to send notifications to players for whom we don't know their coffer/favor values. They would obviously be set once they return to Miscellania, but if a player were to turn these notifications on prior to going back to the island, they could get notifications for "0" coffer/favor values being returned here rather than `null` which would indicate they're not yet known.

Reply:
This is accounted for, because `getLastChanged()` will return `null` until you’ve visited Miscellania, and no message will be sent.